docs/route.md

```diff
@@ -490,9 +490,13 @@ In example 2, since `Layout` is the top-level component in both routes, the DOM
 
 ---
 
-#### Authentication
+#### Redirection
 
-The RouterResolver's `onmatch` hook can be used to run logic before the top level component in a route is initializated. The example below shows how to implement a login wall that prevents users from seeing the `/secret` page unless they login.
+The RouterResolver's `onmatch` hook can be used to run logic before the top level component in a route is initializated. You can use either `m.route.set()` or `window.history.back()` `.forward()`, etc... When redirecting with the `history` API, the `onmatch` hook must return `new Promise(function(){})` to suspend the resolution of the current route. A fresh Promise must be returned every time since recycling a pending Promise leads to memory leaks. `m.route.set()` takes care of cancelling the resolution of the aborted route, there's no need to return a pending Promise in that case.
```

Comment:
It does leak, try doing it in a loop:

```JS
pending = new Promise(_=>_)
while (1) Promise.resolve(pending).then(console.log) // asplode!
```

```
<--- Last few GCs --->

[26862:0x103000000]    62849 ms: Mark-sweep 1329.1 (1385.2) -> 1328.8 (1386.2) MB, 4857.5 / 0.0 ms  allocation failure GC in old space requested
[26862:0x103000000]    67322 ms: Mark-sweep 1328.8 (1386.2) -> 1328.8 (1355.2) MB, 4472.7 / 0.0 ms  last resort gc
[26862:0x103000000]    71993 ms: Mark-sweep 1328.8 (1355.2) -> 1328.8 (1355.2) MB, 4671.3 / 0.0 ms  last resort gc


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x5cf567c0d39 <JS Object>
    2: PromiseAttachCallbacks(aka PromiseAttachCallbacks) [native promise.js:~129] [pc=0x275dcccd5c15](this=0x5cf56704311 <undefined>,r=0x66fb02885a1 <a Promise with map 0x3da59ba3b571>,F=0x279cf17c9c19 <an Object with map 0x3da59ba3b519>,L=0x66fb0274f09 <JS BoundFunction (BoundTargetFunction 0x2187cfb78289)>,M=0x5cf567b0d69 <JS Function PromiseIdRejectHandler (SharedFunctionInfo 0x5cf56750e31)>...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 5: v8::internal::Factory::NewUninitializedFixedArray(int) [/usr/local/bin/node]
 6: v8::internal::(anonymous namespace)::FastElementsAccessor<v8::internal::(anonymous namespace)::FastHoleyObjectElementsAccessor, v8::internal::(anonymous namespace)::ElementsKindTraits<(v8::internal::ElementsKind)3> >::AddArguments(v8::internal::Handle<v8::internal::JSArray>, v8::internal::Handle<v8::internal::FixedArrayBase>, v8::internal::Arguments*, unsigned int, v8::internal::(anonymous namespace)::Where) [/usr/local/bin/node]
 7: v8::internal::(anonymous namespace)::DoArrayPush(v8::internal::Isolate*, v8::internal::BuiltinArguments) [/usr/local/bin/node]
 8: 0x275dccb08506
[1]    26862 abort      node
```

`while (1) Promise.resolve(new Promise(_=>_).then(console.log)` won't crash Node.

The problem comes from the fact that `Promise.resolve(pending)` absorbs `pending`.